Doesn't seem to work with Clang C++20 modules

### Description

I'm not sure if this is a bug in Clang, or in this library. I've been using C++20 modules there but they do have some problems.

https://godbolt.org/z/ndqTvEzGj

I linked to Clang 16.0.0, because as of right now, the trunk is segfaulting. It's been doing that for a few days now. But normally I use Clang trunk when I can. 

### Reproduction steps

Look at the godbolt example, it shows the error. It's a minimal example with no other code in the files.

### Expected vs. actual results

This cryptic compiler error shouldn't be happening.

### Minimal code example

```Shell
https://godbolt.org/z/ndqTvEzGj
```


### Error messages

```Shell
/opt/compiler-explorer/clang-16.0.0/bin/clang++ --gcc-toolchain=/opt/compiler-explorer/gcc-12.2.0   -fcolor-diagnostics -fno-crash-diagnostics -std=c++20 -isystem/opt/compiler-explorer/libs/nlohmann_json/trunk/single_include -g -MD -MT CMakeFiles/mod_json_bug.dir/Mod1_Impl.cc.o -MF CMakeFiles/mod_json_bug.dir/Mod1_Impl.cc.o.d @CMakeFiles/mod_json_bug.dir/Mod1_Impl.cc.o.modmap -o CMakeFiles/mod_json_bug.dir/Mod1_Impl.cc.o -c /app/Mod1_Impl.cc
In file included from /app/Mod1_Impl.cc:2:
/opt/compiler-explorer/libs/nlohmann_json/trunk/single_include/nlohmann/json.hpp:14438:13: error: no matching function for call to 'unescape'
            detail::unescape(reference_token);
            ^~~~~~~~~~~~~~~~
/opt/compiler-explorer/libs/nlohmann_json/trunk/single_include/nlohmann/json.hpp:13782:28: note: in instantiation of member function 'nlohmann::json_pointer<std::basic_string<char>>::split' requested here
        : reference_tokens(split(s))
                           ^
/opt/compiler-explorer/libs/nlohmann_json/trunk/single_include/nlohmann/json.hpp:24402:12: note: in instantiation of member function 'nlohmann::json_pointer<std::basic_string<char>>::json_pointer' requested here
    return nlohmann::json::json_pointer(std::string(s, n));
           ^
```


### Compiler and operating system

linux, clang, or macos clang

### Library version

6af826d0bdb55e4b69e3ad817576745335f243ca via conan

### Validation

- [ ] The bug also occurs if the latest version from the [`develop`](https://github.com/nlohmann/json/tree/develop) branch is used.
- [ ] I can successfully [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests).

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -143,7 +143,7 @@ jobs:
     strategy:
       matrix:
         # older GCC docker images (4, 5, 6) fail to check out code
-        compiler: ['7', '8', '9', '10', '11', '12', '13', '14', 'latest']
+        compiler: ['7', '8', '9', '10', '11', '12', '13', '14', '15', 'latest']
     container: gcc:${{ matrix.compiler }}
     steps:
       - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
@@ -220,6 +220,18 @@ jobs:
         run: cmake -S . -B build -DJSON_CI=On
       - name: Build
         run: cmake --build build --target ci_cuda_example
+
+  ci_module_cpp20:
+    runs-on: ubuntu-latest
+    container: gcc:latest
+    steps:
+      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
+      - name: Get latest CMake and ninja
+        uses: lukka/get-cmake@57c20a23a6cac5b90f31864439996e5b206df9dc # v4.0.1
+      - name: Run CMake
+        run: cmake -S . -B build -DJSON_CI=On
+      - name: Build
+        run: cmake --build build --target ci_module_cpp20
 
   ci_icpc:
     runs-on: ubuntu-latest

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -222,8 +222,11 @@ jobs:
         run: cmake --build build --target ci_cuda_example
 
   ci_module_cpp20:
+    strategy:
+      matrix:
+        container: ['gcc:latest', 'silkeh/clang:latest']
     runs-on: ubuntu-latest
-    container: gcc:latest
+    container: ${{ matrix.container }}
     steps:
       - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
       - name: Get latest CMake and ninja

--- a/cmake/ci.cmake
+++ b/cmake/ci.cmake
@@ -660,6 +660,17 @@ add_custom_target(ci_cuda_example
 )
 
 ###############################################################################
+# C++ 20 modules
+###############################################################################
+
+add_custom_target(ci_module_cpp20
+    COMMAND ${CMAKE_COMMAND}
+        -DCMAKE_BUILD_TYPE=Debug -GNinja
+        -S${PROJECT_SOURCE_DIR}/tests/module_cpp20 -B${PROJECT_BINARY_DIR}/ci_module_cpp20
+    COMMAND ${CMAKE_COMMAND} --build ${PROJECT_BINARY_DIR}/ci_module_cpp20
+)
+
+###############################################################################
 # Intel C++ Compiler
 ###############################################################################
 

--- a/docs/mkdocs/docs/community/quality_assurance.md
+++ b/docs/mkdocs/docs/community/quality_assurance.md
@@ -76,6 +76,7 @@ violations will result in a failed build.
         | GNU 13.3.0                                   | x86_64       | Ubuntu 22.04.1 LTS       | GitHub    |
         | GNU 14.2.0                                   | x86_64       | Ubuntu 22.04.1 LTS       | GitHub    |
         | GNU 14.2.0                                   | arm64        | Linux 6.1.100            | Cirrus CI |
+        | GNU 15.1.0                                   | x86_64       | Ubuntu 22.04.1 LTS       | GitHub    |
         | icpc (ICC) 2021.5.0 20211109                 | x86_64       | Ubuntu 20.04.3 LTS       | GitHub    |
         | MSVC 19.0.24241.7                            | x86          | Windows 8.1              | AppVeyor  |
         | MSVC 19.16.27035.0                           | x86          | Windows-10 (Build 14393) | AppVeyor  |

--- a/include/nlohmann/detail/conversions/from_json.hpp
+++ b/include/nlohmann/detail/conversions/from_json.hpp
@@ -34,6 +34,10 @@
     #include <optional> // optional
 #endif
 
+#if JSON_HAS_FILESYSTEM || JSON_HAS_EXPERIMENTAL_FILESYSTEM
+    #include <string_view> // u8string_view
+#endif
+
 NLOHMANN_JSON_NAMESPACE_BEGIN
 namespace detail
 {

--- a/include/nlohmann/detail/input/binary_reader.hpp
+++ b/include/nlohmann/detail/input/binary_reader.hpp
@@ -53,7 +53,7 @@ enum class cbor_tag_handler_t
 
 @note from https://stackoverflow.com/a/1001328/266378
 */
-static inline bool little_endianness(int num = 1) noexcept
+inline bool little_endianness(int num = 1) noexcept
 {
     return *reinterpret_cast<char*>(&num) == 1;
 }

--- a/include/nlohmann/detail/string_escape.hpp
+++ b/include/nlohmann/detail/string_escape.hpp
@@ -62,7 +62,7 @@ inline StringType escape(StringType s)
  * Note the order of escaping "~1" to "/" and "~0" to "~" is important.
  */
 template<typename StringType>
-static void unescape(StringType& s)
+inline void unescape(StringType& s)
 {
     replace_substring(s, StringType{"~1"}, StringType{"/"});
     replace_substring(s, StringType{"~0"}, StringType{"~"});

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -4819,6 +4819,10 @@ NLOHMANN_JSON_NAMESPACE_END
     #include <optional> // optional
 #endif
 
+#if JSON_HAS_FILESYSTEM || JSON_HAS_EXPERIMENTAL_FILESYSTEM
+    #include <string_view> // u8string_view
+#endif
+
 NLOHMANN_JSON_NAMESPACE_BEGIN
 namespace detail
 {

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -3126,7 +3126,7 @@ inline StringType escape(StringType s)
  * Note the order of escaping "~1" to "/" and "~0" to "~" is important.
  */
 template<typename StringType>
-static void unescape(StringType& s)
+inline void unescape(StringType& s)
 {
     replace_substring(s, StringType{"~1"}, StringType{"/"});
     replace_substring(s, StringType{"~0"}, StringType{"~"});
@@ -9838,7 +9838,7 @@ enum class cbor_tag_handler_t
 
 @note from https://stackoverflow.com/a/1001328/266378
 */
-static inline bool little_endianness(int num = 1) noexcept
+inline bool little_endianness(int num = 1) noexcept
 {
     return *reinterpret_cast<char*>(&num) == 1;
 }

--- a/tests/module_cpp20/CMakeLists.txt
+++ b/tests/module_cpp20/CMakeLists.txt
@@ -1,0 +1,12 @@
+cmake_minimum_required(VERSION 3.28)
+
+project(json_test CXX)
+
+add_executable(json_test)
+
+target_sources(json_test
+    PRIVATE main.cpp
+    PUBLIC FILE_SET cxx_modules TYPE CXX_MODULES FILES json.cpp)
+
+target_compile_features(json_test PUBLIC cxx_std_20)
+target_include_directories(json_test PRIVATE ../../include)

--- a/tests/module_cpp20/json.cpp
+++ b/tests/module_cpp20/json.cpp
@@ -2,7 +2,8 @@ module;
 #include <nlohmann/json.hpp>
 export module json;
 
-export namespace nlohmann {
+export namespace nlohmann
+{
 using ::nlohmann::adl_serializer;
 
 using ::nlohmann::basic_json;

--- a/tests/module_cpp20/json.cpp
+++ b/tests/module_cpp20/json.cpp
@@ -1,0 +1,16 @@
+module;
+#include <nlohmann/json.hpp>
+export module json;
+
+export namespace nlohmann {
+using ::nlohmann::adl_serializer;
+
+using ::nlohmann::basic_json;
+using ::nlohmann::json_pointer;
+
+using ::nlohmann::json;
+using ::nlohmann::ordered_json;
+using ::nlohmann::ordered_map;
+
+using ::nlohmann::json_pointer;
+}  // namespace nlohmann

--- a/tests/module_cpp20/main.cpp
+++ b/tests/module_cpp20/main.cpp
@@ -1,0 +1,5 @@
+import json;
+
+int main() {
+  nlohmann::json j;
+}

--- a/tests/module_cpp20/main.cpp
+++ b/tests/module_cpp20/main.cpp
@@ -1,5 +1,6 @@
 import json;
 
-int main() {
-  nlohmann::json j;
+int main()
+{
+    nlohmann::json j;
 }

--- a/tests/src/unit-regression2.cpp
+++ b/tests/src/unit-regression2.cpp
@@ -396,7 +396,7 @@ struct Example_3810
     Example_3810() = default;
 };
 
-NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(Example_3810, bla); // NOLINT(misc-use-internal-linkage)
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(Example_3810, bla) // NOLINT(misc-use-internal-linkage)
 
 /////////////////////////////////////////////////////////////////////
 // for #4740

--- a/tests/src/unit-user_defined_input.cpp
+++ b/tests/src/unit-user_defined_input.cpp
@@ -60,12 +60,12 @@ TEST_CASE("Custom container member begin/end")
     {
         const char* data;
 
-        const char* begin() const
+        const char* begin() const noexcept
         {
             return data;
         }
 
-        const char* end() const
+        const char* end() const noexcept
         {
             return data + strlen(data); // NOLINT(cppcoreguidelines-pro-bounds-pointer-arithmetic)
         }

